src/collector/skb_tracking/bpf/tracking_hook.bpf.c

```diff
@@ -0,0 +1,144 @@
+ * Indexed in the tracking_config_map by the function ksym address.
+ *
+ * Please keep in sync with its Rust counterpart in collector::skb.
+ */
+struct tracking_config {
+	/* Function is freeing skbs */
+	u8 free;
+	/* Function is invalidating the head of skbs */
+	u8 inv_head;
+};
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, PROBE_MAX);
+	__type(key, u64);
+	__type(value, struct tracking_config);
+} tracking_config_map SEC(".maps");
+
+/* The tracking_info structure stores information on known skbs. It is indexed
+ * in the tracking_map by the skb data address (and in some temporary cases by
+ * the skb address directly).
+ *
+ * In order to uniquely identify skbs, the tuple (addr, timestamp) is used and
+ * must be reported as part of all events (TODO).
+ *
+ * Please keep in sync with its Rust counterpart in collector::skb.
+ */
+struct tracking_info {
+	/* When the skb was first seen */
+	u64 timestamp;
+	/* When the skb was last seen */
+	u64 last_seen;
+	/* Original head address; useful when the head is invalidated */
+	u64 orig_head;
+} __attribute__((packed));
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u64);
+	__type(value, struct tracking_info);
+} tracking_map SEC(".maps");
+
+/* Must be called with a valid skb pointer */
+static __always_inline int track_skb(struct trace_context *ctx,
+				     struct sk_buff *skb)
+{
+	enum skb_drop_reason drop_reason = 0;
+	bool free = false, inv_head = false;
+	struct tracking_info *ti = NULL;
+	struct tracking_config *cfg;
+	u64 head, ksym = ctx->ksym;
+
+	/* Try to retrieve the tracking configuration for this symbol. Only
+	 * specific ones will be found while we want to track skb in all
+	 * functions taking an skb as a parameter. When no tracking
+	 * configuration is found, the function being probed is just quite
+	 * generic.
+	 */
+	cfg = bpf_map_lookup_elem(&tracking_config_map, &ksym);
+	if (cfg) {
+		free = cfg->free;
+		inv_head = cfg->inv_head;
+	}
+
+	head = (u64)BPF_CORE_READ(skb, head);
+	if (!head)
+		return 0;
+
+	ti = bpf_map_lookup_elem(&tracking_map, &head);
+
+	/* No tracking info was found for this skb. */
+	if (!ti) {
+		/* It might be temprorarily stored it using its skb address. */
+		ti = bpf_map_lookup_elem(&tracking_map, (u64 *)&skb);
+		if (ti) {
+			/* If found index it by its data address from now on,
+			 * as others.
+			 */
+			bpf_map_delete_elem(&tracking_map, (u64 *)&skb);
+			bpf_map_update_elem(&tracking_map, &head, ti,
+					    BPF_NOEXIST);
+		}
+	}
+
+	/* Still NULL, this is the first time we see this skb. Create a new
+	 * tracking info.
+	 */
+	if (!ti) {
+		struct tracking_info new = {
+			.timestamp = ctx->timestamp,
+			.last_seen = ctx->timestamp,
+			.orig_head = head,
+		};
+		ti = &new;
+
+		/* No need to globally track it if the first time we see this
+		 * skb is when it is freed.
+		 */
+		if (!free)
+			bpf_map_update_elem(&tracking_map, &head, &new,
+					    BPF_NOEXIST);
```

Comment:
And if (inv_head), we should use skb instead of head right? I know we're doing it later on anyway but why adding it twice?

Reply:
2. could lead to a key leak right?

Reply:
I mean, that's OK since we have the garbage collector but it'll be nice to have those cases documented.

Reply:
If the first event we get is an `inv_head` we add 2 keys: `old_head` and `skb`.
The next event will fail to find a key on `new_head` but it'll find it on `skb`, it'll remove `skb` add `new_head`. Then eventually it'll free `new_head`, but no one removed `old_head`. Maybe I'm missing something?

Reply:
In my original comment what I meant is:
I understand that this mechanism is needed because we need to contemplate situation 2:  we don't know if this packet was a clone and there's actually an skb somewhere with `old_head` so we need to keep it in the map.

Reply:
I was missing the fact that we're tracing head frees not skb ones.